hack/generator/spec/case-study-rolling-storage-versions.md

```diff
@@ -0,0 +1,934 @@
+    dest.Id = person.Id
+    dest.KnownAs = person.KnownAs
+    dest.LegalName = person.LegalName
+
+    // *** Need to check whether we have a mailing address to copy ***
+    if person.MailingAddress != nil {
+        address := &storage.Address{}
+        err := person.MailingAddress.ConvertToStorage(address)
+        if err != nil {
+            return err
+        }
+
+        dest.MailingAddress = address
+    }
+
+    if assignable, ok := person.(AssignableToPerson); ok {
+        err := assignable.AssignTo(dest)
+        if err != nill {
+            return err
+        }
+    }
+
+    return nil
+}
+```
+
+If we instead had an _optional_ field that became _required_ in a later version of the API, the generated code for `ConvertToStorage()` would become simpler as the check for **nil** would not be needed.
+
+## Version Map
+
+![](images/case-study-rolling-storage-2017-07-07.png)
+
+## How often does optionality change?
+
+At the time of writing, there are **100** version-to-version changes where fields became **optional** in the later version of the API, and **99** version-to-version changes where fields became **required**.
+
+# Version 2018-08-08 - Extending nested properties
+
+Defining an address simply as `Street` and `City` has been found to be overly simplistic, so this release makes changes to allow a more flexible approach.
+
+``` go
+package v20180808
+
+type Address struct {
+    // FullAddress shows the entire address as should be used on postage
+    FullAddress  string
+    City         string
+    Country      string
+    PostCode     string
+}
+```
+
+As before, the storage version changes to match, with prior conversions using the property bag to store additional properties:
+
+``` go
+package v20180808storage 
+
+type Address struct {
+    PropertyBag
+    City         *string
+    Country      *string
+    FullAddress  *string
+    PostCode     *string
+}
+```
+
+These changes are entirely similar to those previously covered in version 2014-04-04, above.
+
+## Version Map
+
+In this release, we see that support for both `2014-04-04` and the preview version `2014-04-04preview` has been dropped:
+
+![](images/case-study-rolling-storage-2018-08-08.png)
+
+Users still running earlier releases of the service operator that are using `2014-04-04` or earlier will need to install an intermediate release in order to upgrade to this one.
+
+# Version 2019-09-09 - Changing types
+
+Realizing that some people get deliveries to places that don't appear in any formal database of addresses, in this release the name of the type changes to `Location` and location coordinates are added:
+
+``` go
+package v20190909
+
+type Location struct {
+    FullAddress  string
+    City         string
+    Country      string
+    PostCode     string
+    Lattitude    double
+    Longitide    double
+}
+```
+
+The storage version gets changed in a straightforward way:
+
+``` go
+package v20190909storage 
+
+type Location struct {
```

Comment:
two things;
* shouldn't this include Lattitude and Longitude?
* are properties meant to be alphabetically ordered on the generated storage struct as well as the conversions?

Reply:
Yes to both.

Reply:
Fixed.